### PR TITLE
perf(qwen36): tensor-core router logits + warp top-k + pipelined GDN scan + fused SwiGLU-quant (+24% pp @32k)

### PR DESCRIPTION
**Pull request by widecloud**
## Summary

Four bit-faithful batched-prefill speedups for the Qwen3.6-35B-A3B MoE hybrid: (1) router logits on the bf16 tensor cores instead of a warp-dot GEMV, (2) an in-kernel warp bitonic top-k, (3) the GDN chunk scan's W^/U0 tiles rewritten m-outermost (each `b_m exp(G_m) k[m][d]` formed once per m, not per (i,m)) and the state-decay precomputed per row, and (4) folding the routed-expert SwiGLU + int8 quantize into one kernel (drops the `P×mffn` bf16 round-trip). Each is numerically identical to the path it replaces. Rebased onto current `main` — the GDN changes preserve the short-final-chunk fixes from #604/#608 (store guarded by `i < len`; final gate reads `s_g[len-1]`).

## Proof of speedup

> `main` (`ce6844b`) and this PR built from source into separate trees on the same idle RTX 5090, run **interleaved** (frontier, PR, frontier, PR, …). Median of 3; scored sweep (`SPARKINFER_BENCH_SWEEP_CTXS=0,512,4096,16384,32768 SPARKINFER_BENCH_SWEEP_REPS=1`, `KV_INT8=1`). The bench binary is used instead of `bench.sh` because `bench.sh` is single-context and defaults to a different model.

- [x] Tested on **RTX 5090** (`sm_120`)

**Decode tok/s** (this PR does **not** target decode; shown as a no-regression check):

| | decode tok/s |
|---|--:|
| before (main) | 480.9 |
| after (this PR) | 482.6 |

**Prefill pp tok/s** (Qwen3.6-35B-A3B-UD-Q4_K_M, best context = 32768):

| | prefill pp tok/s |
|---|--:|
| before prefill (main) | 18788.4 |
| after prefill (this PR) | 23309.1 |

All scored Qwen3.6 contexts, median of 3 (marginal over frontier):

| ctx | main | this PR | marginal |
|---|--:|--:|--:|
| 512 | 5804.8 | 5985.5 | +3.1% |
| 4096 | 14730.3 | 17100.5 | +16.1% |
| 16384 | 18543.7 | 22429.9 | +21.0% |
| 32768 | 18788.4 | 23309.1 | **+24.1%** |

Qwen3.5/Qwythos prefill (cross-model guard — the GDN scan is shared, so it also speeds up here; no regression):

| ctx | main | this PR | Δ |
|---|--:|--:|--:|
| 4096 | 19728.1 | 20514.2 | +4.0% |
| 32768 | 19701.6 | 20384.5 | +3.5% |
| 65536 | 19603.2 | 20260.7 | +3.4% |
| 131072 | 19083.2 | 19696.3 | +3.2% |

```text
# same idle box, separate source builds, rep 1 of 3 (raw SWEEP_JSON, prefill_pp):
# Qwen3.6:
main:    {"512":5803.32,"4096":14784.72,"16384":18587.81,"32768":18849.20}
this PR: {"512":5985.53,"4096":17107.26,"16384":22563.19,"32768":23414.84}
# Qwythos guard:
main:    {"4096":19893.39,"32768":19844.01,"65536":19720.02,"131072":19152.32}
this PR: {"4096":20551.04,"32768":20430.27,"65536":20297.01,"131072":19735.72}
```

## Correctness

Every changed kernel is a numerically-faithful replacement of the path it supersedes (same inputs, same accumulate order → bit-identical up to the down-scatter `atomicAdd` nondeterminism already present on `main`). Checked with `qwen3_gguf_prefill_check` (batched prefill vs the per-token reference, 64 tail positions), including a **non-32-multiple prompt length (300)** to exercise the short-final-chunk GDN path:

```text
# prefix / KL(main) / KL(this PR):
300  : 0.01213 / 0.00944   (no IMA — short-final-chunk store guard preserved)
512  : 0.00745 / 0.00682
4096 : 0.00720 / 0.00677
8320 : 0.01680 / 0.01524
```

KL stays within `main`'s own run-to-run band at every length, and prefix=300 runs clean (the #604/#608 short-final-chunk OOB/overflow fixes are retained after the rebase).

### Implementation

- `kernels/csrc/cuda/fused/prefill_router_mma.cu` (+ `prefill_router_mma.h`): bf16 tensor-core router-logits GEMM; `launch_pfm_router_logits_mma` returns false (caller falls back to the reference dot) when disabled or the shape is not tile-aligned.
- `kernels/csrc/cuda/moe/router.cu`: in-kernel warp bitonic top-k.
- `kernels/csrc/cuda/fused/prefill_gdn_chunk.cu`: m-outermost W^/U0 tiles + per-row state-decay precompute; stores guarded by `i < len` and the final gate reads `s_g[len-1]` (keeps #604/#608).
- `runtime/src/models/qwen35_prefill.cpp`: route logits through the mma path (with fallback); replace `swiglu`+`quantize_rows_i8` with the fused `swiglu_quant_i8` across the MoE branches.

### CODE REVIEW

**Comment by skyrocket2026** [comment]
## ✅ sparkinfer auto-eval — `1f0e02f`

| metric | value |
|---|---|
| **label** | `eval:XL` |
| Qwen3.5 score | `eval-qwen35:S` (pass) |
| Qwen3.6 score | `eval-qwen36:XL` (pass) |
| Qwen3.5 vs same-box main | 18545.91 tok/s → +5.5% (+1073.7) |
| Qwen3.5 scored decode (128 ctx · 128-context) | 254.78 tok/s |
| Qwen3.5 scored prefill (32768 ctx · 32k-context) | 19619.64 pp tok/s · TTFT −5.5% · `eval-prefill:S` |
| Qwen3.5 CB mixed TTFT no-regression gate | 0.487s vs main 0.504s · TTFT −3.4% · pass |
| Qwen3.5 correctness | top-1 96.0% · KL 0.0256 |
| Qwen3.5 128-token no-regression gate | 254.78 tok/s vs main 254.92 tok/s · pass |
| Qwen3.5 4k-context no-regression gate | 246.21 tok/s vs main 246.35 tok/s · pass |
| Qwen3.5 32k-context no-regression gate | 244.39 tok/s vs main 244.65 tok/s · pass |
| Qwen3.5 64k-context no-regression gate | 244.31 tok/s vs main 244.52 tok/s · pass |
| Qwen3.5 128k-context no-regression gate | 243.48 tok/s · pass |
| Qwen3.5 4k prefill no-regression gate | 18538.94 pp tok/s vs main 17555.97 pp tok/s · pass |
| Qwen3.5 32k prefill no-regression gate | 19619.64 pp tok/s vs main 18545.91 pp tok/s · pass |
| Qwen3.5 64k prefill no-regression gate | 19205.23 pp tok/s vs main 18210.6 pp tok/s · pass |
| Qwen3.5 128k prefill no-regression gate | 18453.81 pp tok/s vs main 17539.21 pp tok/s · pass |
| Qwen3.6 vs same-box main | 17279.17 tok/s → +19.3% (+4126.8) |
| Qwen3.6 scored decode (128 ctx · 128-context) | 425.85 tok/s |
| Qwen3.6 scored prefill (32768 ctx · 32k-context) | 21405.99 pp tok/s · TTFT −19.3% · `eval-prefill:XL` |
| Qwen3.6 CB mixed TTFT no-regression gate | 0.475s vs main 0.568s · TTFT −16.4% · pass |
| Qwen3.6 correctness | top-1 92.9% · KL 0.0511 |
| Qwen3.6 128-token no-regression gate | 425.85 tok/s vs main 425.89 tok/s · pass |
| Qwen3.6 512-context no-regression gate | 419.36 tok/s vs main 419.42 tok/s · pass |
| Qwen3.6 4k-context no-regression gate | 402.06 tok/s vs main 402.11 tok/s · pass |
| Qwen3.6 16k-context no-regression gate | 388.0 tok/s vs main 388.18 tok/s · pass |
| Qwen3.6 32k-context no-regression gate | 364.19 tok/s vs main 364.5 tok/s · pass |
| Qwen3.6 128 prefill no-regression gate | 0.0 pp tok/s · pass |
| Qwen3.6 512 prefill no-regression gate | 4908.03 pp tok/s vs main 4779.94 pp tok/s · pass |
| Qwen3.6 4k prefill no-regression gate | 15136.05 pp tok/s vs main 12939.52 pp tok/s · pass |
| Qwen3.6 16k prefill no-regression gate | 20418.46 pp tok/s vs main 16683.57 pp tok/s · pass |
| Qwen3.6 32k prefill no-regression gate | 21405.99 pp tok/s vs main 17279.17 pp tok/s · pass |
| **Qwen3.5 optimize** | `eval:S` · 19619.64 tok/s · pass |
| Qwen3.5 optimize — Qwen3.6-35B-A3B guard accuracy | top-1 92.9% · KL 0.0511 · pass |
| Qwen3.5 optimize — Qwythos-9B (Q4_K_M) 128 | 254.78 tok/s · pass |
| Qwen3.5 optimize — Qwythos-9B (Q4_K_M) 4k | 246.21 tok/s · pass |
| Qwen3.5 optimize — Qwythos-9B (Q4_K_M) 32k | 244.39 tok/s · pass |
| Qwen3.5 optimize — Qwythos-9B (Q4_K_M) 64k | 244.31 tok/s · pass |
| Qwen3.5 optimize — Qwythos-9B (Q4_K_M) 128k | 243.48 tok/s · pass |
| **Qwen3.6 optimize** | `eval:XL` · 21405.99 tok/s · pass |
| Qwen3.6 optimize — Qwythos-9B (Q4_K_M) guard accuracy | top-1 96.0% · KL 0.0256 · pass |
| Qwen3.6 optimize — Qwen3.6-35B-A3B 128 | 425.85 tok/s · pass |
| Qwen3.6 optimize — Qwen3.6-35B-A3B 512 | 419.36 tok/s · pass |
| Qwen3.6 optimize — Qwen3.6-35B-A3B 4k | 402.06 tok/s · pass |
| Qwen3.6 optimize — Qwen3.6-35B-A3B 16k | 388.0 tok/s · pass |
| Qwen3.6 optimize — Qwen3.6-35B-A3B 32k | 364.19 tok/s · pass |

Verified speedup over same-box origin/main — 21405.99 tok/s (main was 17279.17 tok/s).

_RTX 5090 (sm_120) · 128/512/4k/16k/32k guarded · Qwen3.5 prefill at 4k/32k/64k/128k · Qwen3.6 prefill at 128/512/4k/16k/32k · scored vs same-box main · built from source · correctness vs llama.cpp. Automated — **not merged**; merge manually after review._
